include/ginkgo/core/distributed/partition.hpp

```diff
@@ -0,0 +1,321 @@
+     * Returns the part IDs of the ranges in this partition.
+     * For each range from get_range_bounds(), it stores the part ID in the
+     * interval [0, get_num_parts() - 1].
+     *
+     * @return  part ID array.
+     */
+    const comm_index_type* get_part_ids() const noexcept
+    {
+        return part_ids_.get_const_data();
+    }
+
+    /**
+     * Returns the part-local starting index for each range in this partition.
+     *
+     * Consider the partition on `[0, 10)` with
+     * ```
+     * p_1 = [0-4) + [7-10),
+     * p_2 = [4-7).
+     * ```
+     * Then `range_starting_indices[0] = 0`, `range_starting_indices[1] = 0`,
+     * `range_starting_indices[2] = 4`.
+     *
+     * @return  part-local starting index array.
+     */
+    const local_index_type* get_range_starting_indices() const noexcept
+    {
+        return starting_indices_.get_const_data();
+    }
+
+    /**
+     * Returns the part size array.
+     * part_sizes[p] stores the total number of indices in part `p`.
+     *
+     * @return  part size array.
+     */
+    const local_index_type* get_part_sizes() const noexcept
+    {
+        return part_sizes_.get_const_data();
+    }
+
+    /**
+     * Returns the size of a part given by its part ID.
+     * @warning Triggers a copy from device to host.
+     *
+     * @param part  the part ID.
+     *
+     * @return  size of part.
+     */
+    local_index_type get_part_size(comm_index_type part) const
+    {
+        return this->get_executor()->copy_val_to_host(
+            part_sizes_.get_const_data() + part);
+    }
+
+    /**
+     * Checks if each part has no more than one contiguous range.
+     *
+     * @return  true if each part has no more than one contiguous range.
+     */
+    bool has_connected_parts();
+
+    /**
+     * Checks if the ranges are ordered by their part index.
+     *
+     * Implies that the partition is connected.
+     *
+     * @return  true if the ranges are ordered by their part index.
+     */
+    bool has_ordered_parts();
+
+    /**
+     * Builds a partition from a given mapping global_index -> part_id.
+     *
+     * @param exec  the Executor on which the partition should be built
+     * @param mapping  the mapping from global indices to part IDs.
+     * @param num_parts  the number of parts used in the mapping.
+     *
+     * @return  a Partition representing the given mapping as a set of ranges
+     */
+    static std::unique_ptr<Partition> build_from_mapping(
+        std::shared_ptr<const Executor> exec,
+        const Array<comm_index_type>& mapping, comm_index_type num_parts);
+
+    /**
+     * Builds a partition consisting of contiguous ranges, one for each part.
+     *
+     * @param exec  the Executor on which the partition should be built
+     * @param ranges  the boundaries of the ranges representing each part.
+                      Part i contains the indices [ranges[i], ranges[i + 1]).
+                      Has to contain at least one element.
```

Comment:
I guess here? I can not get the part how to handle one element. Do I miss anything?

Reply:
You're right, I was thinking about something else. To clarify, I added the comment to prevent issues from using `size(ranges) - 1`. If you pass just one element you will still get an empty range.

Reply:
I see. that makes sense